autometa/taxonomy/ncbi.py

```diff
@@ -362,62 +370,88 @@ def parse_merged(self):
-        f"{args.query_taxid} Lineage:\n{ncbi.lineage(taxid=args.query_taxid)}\n"
-        f"{args.query_taxid} is_common_ancestor {query_taxid_parent}: "
-        f"{ncbi.is_common_ancestor(parent_taxid=query_taxid_parent, child_taxid=args.query_taxid)}\n"
-    )
+        lineage_a_taxids = {ancestor.get("taxid") for ancestor in self.lineage(taxid_A)}
+        lineage_b_taxids = {ancestor.get("taxid") for ancestor in self.lineage(taxid_B)}
+        common_ancestor = lineage_b_taxids.intersection(lineage_a_taxids)
+        common_ancestor.discard(1)  # This discards root
+        return True if common_ancestor else False
+
+    def convert_taxid_dtype(self, taxid):
+        """
+        Converts the given `taxid` to integer. It also checks if the given `taxid` is a positive integer and
+        if is is present in either nodes.dmp or names.dmp or merged.dmp.
+
+        Parameters
+        ----------
+        taxid : int
+            identifer for a taxon in NCBI taxonomy databases - nodes.dmp, names.dmp or merged.dmp
+
+        Returns
+        -------
+        int
+            `taxid` if the `taxid` is a positive integer and present in either nodes.dmp or names.dmp or
+            taxid recovered from merged.dmp
+
+        Raises
+        ------
+        ValueError
+            Provided `taxid` is not a positive integer
+        DatabaseOutOfSyncError
+            NCBI databases nodes.dmp, names.dmp and merged.dmp are out of sync with each other
+        """
+        # Checking taxid instance format
+        # This checks if an integer has been added as str, eg. "562"
+        if isinstance(taxid, str):
+            invalid_chars = set(string.punctuation + string.ascii_letters)
+            invalid_chars.discard(".")
+            if {char for char in taxid if char in invalid_chars} or taxid.count(
+                "."
+            ) > 1:
+                raise ValueError(f"taxid contains invalid character(s)! Given: {taxid}")
+            taxid = float(taxid)
+        # a boolean check is needed as they will evaluate silently to 0 or 1 when cast as ints. FALSE=0, TRUE=1
+        # float(taxid).is_integer() checks if it is something like 12.0 vs. 12.9
+        # is_integer only takes float as input else raises error, thus isinstance( ,float) is used before it to make sure a float is being passed
+        if isinstance(taxid, bool) or (
+            isinstance(taxid, float) and not taxid.is_integer()
+        ):
+            raise ValueError(f"taxid must be an integer! Given: {taxid}")
+        taxid = int(taxid)
+        if taxid <= 0:
+            raise ValueError(f"Taxid must be a positive integer! Given: {taxid}")
+        # Checking databases
+        if taxid not in self.names and taxid not in self.nodes:
+            if taxid not in self.merged:
+                err_message = f"Databases out of sync. {taxid} not in found in nodes.dmp, names.dmp or merged.dmp."
+                logger.warning(err_message)
```

Comment:
I do not think this should be logged here, but rather in the functions that are calling `self.convert_taxid_dtype`.

Reply:
I am not sure what would be the advantage to that. I think logging the error here is good because:
1. It is easy to distinguish the two different errors and thus log two different error messages. If we log the error in the client function we would not be able to log two different error messages and provide less specificity about the error to the user.
2. I think it saves a couple of lines of code. As we just need to write `logger.warning(err_message)` once and not repeat it again and again for every client function.

I could be wrong so please let me know what you think about this.

Reply:
By only logging when the function is called, you provide the user the opportunity to choose to log the error or not. By placing the log message here, the user no longer has the option to emit the log message. Since you are already raising the error, the user would still have the option to log the message when error handling.

>1. It is easy to distinguish the two different errors and thus log two different error messages. If we log the error in the client function we would not be able to log two different error messages and provide less specificity about the error to the user.

I'm not sure what you mean here. Both `logger.warning` and `DatabaseOutOfSyncError` emit the same message. 

>2. I think it saves a couple of lines of code. As we just need to write logger.warning(err_message) once and not repeat it again and again for every client function.

Removing control from the developer in the name of saving lines of code does not seem like a fair trade. This would only need to be logged in every client function if every client function would like to log the message. This will be completely dependent on the error handling specific for each client function.